src/index.ts

```diff
@@ -2,92 +2,126 @@ import cp = require('child_process')
 import fs = require('fs')
 import p = require('path')
 
-// Logger
-const l = (msg: string): void => console.log(`husky - ${msg}`)
-
-// Git command
-const git = (args: string[]): cp.SpawnSyncReturns<Buffer> =>
-  cp.spawnSync('git', args, { stdio: 'inherit' })
-
-export function install(dir = '.husky'): void {
-  if (process.env.HUSKY === '0') {
-    l('HUSKY env variable is set to 0, skipping install')
-    return
-  }
-
-  // Ensure that we're inside a git repository
-  // If git command is not found, status is null and we should return.
-  // That's why status value needs to be checked explicitly.
-  if (git(['rev-parse']).status !== 0) {
-    return
-  }
-
-  // Custom dir help
-  const url = 'https://typicode.github.io/husky/#/?id=custom-directory'
-
-  // Ensure that we're not trying to install outside of cwd
-  if (!p.resolve(process.cwd(), dir).startsWith(process.cwd())) {
-    throw new Error(`.. not allowed (see ${url})`)
-  }
-
-  // Ensure that cwd is git top level
-  if (!fs.existsSync('.git')) {
-    throw new Error(`.git can't be found (see ${url})`)
-  }
-
-  try {
-    // Create .husky/_
-    fs.mkdirSync(p.join(dir, '_'), { recursive: true })
-
-    // Create .husky/_/.gitignore
-    fs.writeFileSync(p.join(dir, '_/.gitignore'), '*')
+interface Options {
+  log: (msg: string) => void
+  error: (msg: string) => void
+}
 
-    // Copy husky.sh to .husky/_/husky.sh
-    fs.copyFileSync(p.join(__dirname, '../husky.sh'), p.join(dir, '_/husky.sh'))
+interface CustomOptions {
+  log?: (msg: string) => void
+  error?: (msg: string) => void
+}
 
-    // Configure repo
-    const { error } = git(['config', 'core.hooksPath', dir])
-    if (error) {
-      throw error
-    }
-  } catch (e) {
-    l('Git hooks failed to install')
-    throw e
-  }
+interface Husky {
+  install: (dir: string) => void
+  set: (file: string, cmd: string) => void
+  add: (file: string, cmd: string) => void
+  uninstall: () => void
+}
 
-  l('Git hooks installed')
+// Default options
+const defaultOptions: Options = {
+  log: (msg: string): void => console.log(`husky - ${msg}`),
+  error: (msg: string): void => console.error(`husky - ${msg}`),
 }
 
-export function set(file: string, cmd: string): void {
-  const dir = p.dirname(file)
-  if (!fs.existsSync(dir)) {
-    throw new Error(
-      `can't create hook, ${dir} directory doesn't exist (try running husky install)`,
-    )
-  }
+// Git command
+const git = (args: string[]): cp.SpawnSyncReturns<Buffer> =>
+  cp.spawnSync('git', args, { stdio: 'inherit' })
 
-  fs.writeFileSync(
-    file,
-    `#!/usr/bin/env sh
+const defaultHusky = configure()
+
+export function configure(customOptions: CustomOptions = {}): Husky {
+  const options: Options = { ...defaultOptions, ...customOptions }
```

Comment:
Here's the 4th one :)

Sorry I am not that experienced with github's review flow

Reply:
Me neither :smile:.

I agree that the current naming is quite odd and your proposal is more readable.
Let's change that.

Reply:
I've just embedded the log functions in a `logger` section within `Options` which is clearly more readable.

The drawback from a TypeScript point of view is that one has to define both `log` and `error` functions if the `logger` is overridden (I have checked that TypeScript will raise an error if `logger` is partially overridden).
This seems ok as it would be odd for someone to only partially redefine the logger.

What I am more concerned about is people using husky in pure javascript.
As they do not have the TypeScript safety net, they might forgot to override all `logger` functions and this will fail badly at runtime.

So I am wondering if I should not make all the logger functions optional in order to avoid this situation...

Reply:
> What I am more concerned about is people using husky in pure javascript.
> As they do not have the TypeScript safety net, they might forgot to override all logger functions and this will fail badly at runtime.

I just refactored the logger initialization in order to avoid this type of issue.